simpleline/render/widget_containers.py

```diff
@@ -0,0 +1,349 @@
+        """Get a widget container.
+
+        :param container_id: ID of the container in the list.
+        :type container_id: int
+
+        :returns: WidgetContainer class.
+        :raise: IndexError
+        """
+        return self._widgets[container_id]
+
+    def modify_prompt(self, prompt):
+        """Add user options for items in the list to prompt.
+
+        :param prompt: Prompt for modification.
+        :type prompt: Class based on `simpleline.render.prompt.Prompt`.
+        """
+        key_pattern = self._key_pattern
+        if key_pattern:
+            (key_hint, description) = key_pattern.get_prompt_help_and_description(len(self._widgets) - 1)
+            prompt.add_option(key_hint, description)
+
+    def process_user_input(self, key):
+        """Process input from the user if any of the items in the list was called.
+
+        This method must be called in `UIScreen.input()` method if list widget should call the callbacks.
+
+        :param key: Key pressed from user.
+        :type key: str
+
+        :returns: Nothing.
+        """
+        if not self._key_pattern or type(key) != str:
+            return False
+
+        res = self._key_pattern.get_widget_id_from_key(key)
+        if res is not None:
+            try:
+                container = self._widgets[res]
+                container.callback(container.data)
+                return True
+            except IndexError:  # container widget with this id doesn't exists
+                return False
+
+        return False
+
+    def render(self, width):
+        """Render widgets to it's internal buffer.
+
+        :param width: the maximum width the widget can use
+        :type width: int
+
+        :return: nothing
+        """
+        super().render(width)
+
+        ordered_map = self._get_ordered_map()
+        lines_per_rows = self.render_and_calculate_lines_per_rows(ordered_map)
+
+        # the leftmost empty column
+        col_pos = 0
+
+        for col in ordered_map:
+            row_pos = 0
+
+            # render and draw contents of column
+            for row_id, item_id in enumerate(col):
+                container = self._widgets[item_id]
+                widget = container.widget
+
+                # set cursor to first line and leftmost empty column
+                self.set_cursor_position(row_pos, col_pos)
+
+                if self._key_pattern is not None:
+                    number_widget = self._numbering_widgets[item_id]
+                    widget_width = len(number_widget.text)
+                    self.draw(number_widget)
+                    self.set_cursor_position(row_pos, col_pos + widget_width)
+
+                self.draw(widget, block=True)
+                row_pos = row_pos + lines_per_rows[row_id]
+
+            # recompute the leftmost empty column
+            col_pos = max((col_pos + self._columns_width), self.width) + self._spacing
+
+    def render_and_calculate_lines_per_rows(self, ordered_items):
```

Comment:
Nope it is here on purpose. The problem here is that items must be rendered before the lines could be calculated. So this method is giving required order of calls. I'll rework this to be better understandable and reduce the number of methods.